Add WebUi\Address config option

fmatthew5876 opened the pull request:
As I mentioned in a previous PR, I had an issue with qbittorrent-nox failing to `listen()` on my FreeBSD jail. The reason it was failing is because the webui is hard coded to use `QHostAddress::Any` which tries to bind to IPv4 and IPv6. This caused the failure because my jail has IPv6 disable.

This patch adds a "WebUI\Address" config option. It uses the string parsing built into `QHostAddress`. One caveat is that there is no equivalent string for `QHostAddress::Any`, so I've opted to use the string "*" to mean Any and this is the default.

Since there is parsing now being done in `Preferences`, I've had to add a `PreferencesParsingError` exception class.

Chocobo1 commented on src/base/preferences.cpp:
```diff
@@ -463,6 +463,28 @@ void Preferences::setServerDomains(const QString &str)
 
+QHostAddress Preferences::getWebUiAddress() const
+{
+    auto addr_s = value("Preferences/WebUI/Address", "*").toString();
```
this project use camelcase for variables, please fix coding style

Chocobo1 commented on src/base/preferences.cpp:
```diff
@@ -463,6 +463,28 @@ void Preferences::setServerDomains(const QString &str)
+QHostAddress Preferences::getWebUiAddress() const
+{
+    auto addr_s = value("Preferences/WebUI/Address", "*").toString();
+    if(addr_s == "*") {
```
`if (`

thalieht replied:
In other places too. 
Also another unrelated comment that doesn't deserve it's own line because there is no rule for it IIRC, is that reference and pointer symbols are usually attached to the variable name in this project.

Chocobo1 commented on src/base/preferences.cpp:
```diff
@@ -463,6 +463,28 @@ void Preferences::setServerDomains(const QString &str)
+    if(!addr.setAddress(addr_s)) {
+        throw PreferencesParsingError(tr("Failed to parse IP Address string `%1'").arg(addr_s));
+    }
+    return addr;
```
I would just `return  QHostAddress(addr_s);` and let `m_httpServer->listen` tell whether there are any error.

Chocobo1 commented on src/webui/webui.cpp:
```diff
@@ -93,7 +94,7 @@ void WebUI::init()
 
         if (!m_httpServer->isListening()) {
-            bool success = m_httpServer->listen(QHostAddress::Any, m_port);
+            bool success = m_httpServer->listen(address, m_port);
```
just inline the call

fmatthew5876 replied:
I put it early so that the exception would be thrown as early as possible to report the error. But now if we just returning an empty QHostAddress on error this makes sense.

Chocobo1 commented:
You need to add the respective option in GUI under the WebUI tab.

thalieht commented on src/base/preferences.h:
```diff
@@ -80,6 +81,13 @@ namespace DNS
 class SettingsStorage;
 
+class PreferencesParsingError : public std::runtime_error {
+    public:
```
Should not be indented. Also the opening bracket above should be in it's own line.

thalieht commented on src/webui/webui.cpp:
```diff
@@ -54,6 +54,7 @@ void WebUI::init()
 
     const quint16 oldPort = m_port;
     m_port = pref->getWebUiPort();
+    const auto address = pref->getWebUiAddress();
```
I don't think it's obvious here that `getWebUiAddress()` returns a `QHostAddress`

fmatthew5876 commented:
Rebased the commit.

* Added Qt Gui Support
* Added WebUI HTML UI support
* Fixed style issues

It looks like the preferences system is highly wedded to `QVariant`. That and also the json code used for webui. For that reason I've changed the WebUiAddress to be a `QString`, and I do the parsing within `webui.cpp`.

Chocobo1 commented on src/gui/optionsdlg.ui:
```diff
@@ -2782,6 +2782,22 @@ Use ';' to split multiple entries. Can use wildcard '*'.</string>
+                 <item>
+                  <widget class="QLineEdit" name="textWebUiAddress">
+                   <property name="toolTip">
+                    <string>Ip Address which the WebUi will bind to.
```
* Ip Address -> IP address
* WebUi -> Web UI

thalieht replied:
This doesn't seem right. Maybe
"The IP address that the Web UI will bind to."?

Chocobo1 commented on src/base/preferences.cpp:
```diff
@@ -463,6 +463,16 @@ void Preferences::setServerDomains(const QString &str)
     setValue("Preferences/WebUI/ServerDomains", str);
 }
 
+QString Preferences::getWebUiAddress() const
```
@glassez 
Should it return `QString` or `QHostAddress`?

glassez replied:
I would leave string here.

fmatthew5876 replied:
I chose `QString` as `QHostAddress` isn't supported by `QVariant`.

Chocobo1 replied:
OK since it's better for webUI serializing.

Chocobo1 commented on src/base/preferences.cpp:
```diff
@@ -1475,3 +1485,4 @@ void Preferences::apply()
     if (SettingsStorage::instance()->save())
         emit changed();
 }
+
```
this empty line is unnecessary

glassez commented on src/webui/webui.cpp:
```diff
@@ -93,7 +93,16 @@ void WebUI::init()
 
         if (!m_httpServer->isListening()) {
-            bool success = m_httpServer->listen(QHostAddress::Any, m_port);
+			const QString addressString = pref->getWebUiAddress();
```
Wrong indentation in these lines.

glassez replied:
Perhaps, the author forgot to consider project coding guidelines and use tabs instead of spaces for indentation... There are also some other coding style breaks in this PR.

thalieht commented on src/webui/webui.cpp:
```diff
@@ -93,7 +93,16 @@ void WebUI::init()
+			}
+			else {
+				address.setAddress(addressString);
+			}
```
Something's wrong with the indentation of these lines up to line 96.

glassez commented on src/base/preferences.cpp:
```diff
@@ -463,6 +463,16 @@ void Preferences::setServerDomains(const QString &str)
+    return value("Preferences/WebUI/Address", "*").toString();
+}
+
+void Preferences::setWebUiAddress(const QString& addr)
```
Incorrect coding style.

glassez commented on src/base/preferences.cpp:
```diff
@@ -463,6 +463,16 @@ void Preferences::setServerDomains(const QString &str)
 
+QString Preferences::getWebUiAddress() const
+{
+    return value("Preferences/WebUI/Address", "*").toString();
```
It's better to return trimmed string here.

glassez commented on src/base/preferences.cpp:
```diff
@@ -463,6 +463,16 @@ void Preferences::setServerDomains(const QString &str)
+
+void Preferences::setWebUiAddress(const QString& addr)
+{
+    setValue("Preferences/WebUI/Address", addr);
```
I would use `addr.trimmed()`.

glassez commented on src/base/preferences.h:
```diff
@@ -174,6 +174,8 @@ class Preferences: public QObject
     QString getServerDomains() const;
     void setServerDomains(const QString &str);
+    QString getWebUiAddress() const;
+    void setWebUiAddress(const QString& addr);
```
Incorrect coding style.

thalieht replied:
I'm always reluctant to point that out because there is no rule for it.

fmatthew5876 commented:
This is what happens when you make a PR at 2am. Sorry for all of the style violations again.

Should be fixed now. Also we need a clang-format file!

glassez commented on src/webui/webui.cpp:
```diff
@@ -93,7 +93,15 @@ void WebUI::init()
-            bool success = m_httpServer->listen(QHostAddress::Any, m_port);
+            const QString addressString = pref->getWebUiAddress();
+            QHostAddress address;
+            if(addressString == "*") {
```
Missing space after keyword `if`

glassez commented on src/webui/webui.cpp:
```diff
@@ -93,7 +93,15 @@ void WebUI::init()
+            QHostAddress address;
+            if(addressString == "*") {
+                address = QHostAddress::Any;
+            } else {
```
`else` should be on the next line

Chocobo1 commented on src/webui/webui.cpp:
```diff
@@ -93,7 +93,15 @@ void WebUI::init()
         if (!m_httpServer->isListening()) {
-            bool success = m_httpServer->listen(QHostAddress::Any, m_port);
+            const QString addressString = pref->getWebUiAddress();
+            QHostAddress address;
```
I would write:
```c++
const QHostAddress address = (addressString == "*")
  ? QHostAddress::Any : QHostAddress(addressString);
```

fmatthew5876 replied:
Ok sure, but how about using `auto` :)

Chocobo1 replied:
IMO it's ok to use it in this case.

Chocobo1 commented on src/gui/optionsdlg.ui:
```diff
@@ -2782,6 +2782,22 @@ Use ';' to split multiple entries. Can use wildcard '*'.</string>
+                 <item>
+                  <widget class="QLineEdit" name="textWebUiAddress">
+                   <property name="toolTip">
+                    <string>IP Address that the Web UI will bind to.
```
I don't think "Address" should be capitalized.

Chocobo1 replied:
Also please move this QHBoxLayout (`horizontalLayout_2`) above "Server domain" layout and remove the spacer, as so:
![pic](https://user-images.githubusercontent.com/9395168/31557458-590aff64-b07b-11e7-8c84-b4a2d58a563a.png)

glassez commented:
>One caveat is that there is no equivalent string for QHostAddress::Any, so I've opted to use the string "*" to mean Any and this is the default.

I would use empty string in case of Any address. @Chocobo1, what do you say?

Chocobo1 commented:
>I would use empty string in case of Any address. @Chocobo1, what do you say?

Why not both? I don't have a preference about that, yet I think we should treat the empty string case as "Any address".

Chocobo1 commented on src/webui/webui.cpp:
```diff
@@ -93,7 +93,11 @@ void WebUI::init()
         if (!m_httpServer->isListening()) {
-            bool success = m_httpServer->listen(QHostAddress::Any, m_port);
+            const QString addressString = pref->getWebUiAddress();
+            const auto address = (addressString == "*" || addressString == "")
```
`addressString.isEmpty()`

fmatthew5876 replied:
wow you are super fast!

Chocobo1 replied:
I just happen to be online ;)

fmatthew5876 commented:
Ok, this is ready for review again. I also fixed the webui options dialog to match the changes to the QT one.

Chocobo1 commented on src/gui/optionsdlg.cpp:
```diff
@@ -648,6 +649,7 @@ void OptionsDialog::saveOptions()
     pref->setWebUiEnabled(isWebUiEnabled());
     if (isWebUiEnabled()) {
         pref->setServerDomains(m_ui->textServerDomains->text());
+        pref->setWebUiAddress(webUiAddress());
```
You can just inline it.

ps. Everything else is fine.

fmatthew5876 replied:
The line below calls `webUiPort()` so I copied that idiom. If one is changed so should the other no?

Let me know what is preferred here.

Chocobo1 replied:
>The line below calls webUiPort() so I copied that idiom. If one is changed so should the other no?

Yes, if you're willing to refactor it.

>Let me know what is preferred here.

If other devs don't object inlining it, let's wait a bit.

glassez replied:
There is no reason for having additional getters for it.

fmatthew5876 replied:
Ok I pushed this as a second commit.

glassez commented on src/webui/webui.cpp:
```diff
@@ -93,7 +93,11 @@ void WebUI::init()
-            bool success = m_httpServer->listen(QHostAddress::Any, m_port);
+            const QString addressString = pref->getWebUiAddress();
+            const auto address = (addressString == "*" || addressString.isEmpty())
+              ? QHostAddress::Any : QHostAddress(addressString);
```
Wrong indentation.

fmatthew5876 replied:
I'm assuming you mean the `?`. This was followed from the example given above. Anyway, I've indented it to 4 lines.

glassez replied:
The examples that we give to each other here, it's not a source for blind copy. We do this often in a hurry. We must take this just as a hint (although I agree, some are pretty finished look).

fmatthew5876 commented:
Rebased again

glassez commented:
>Rebased again

You're done it incorrectly. Consider your current history.

fmatthew5876 commented:
Rebased it against latest upstream master. Looks ok to me. Or am I missing something?

Chocobo1 commented on src/gui/optionsdlg.cpp:
```diff
@@ -648,7 +649,8 @@ void OptionsDialog::saveOptions()
     if (isWebUiEnabled()) {
         pref->setServerDomains(m_ui->textServerDomains->text());
-        pref->setWebUiPort(webUiPort());
+        pref->setWebUiAddress(m_ui->textWebUiAddress->text());
```
@fmatthew5876 
You are missing this:
`m_ui->textWebUiAddress->setText(pref->getWebUiAddress());`

fmatthew5876 replied:
Fixed in latest rebase

evsh commented on src/base/preferences.cpp:
```diff
@@ -463,6 +463,16 @@ void Preferences::setServerDomains(const QString &str)
 
+QString Preferences::getWebUiAddress() const
+{
+    return value("Preferences/WebUI/Address", "*").toString().trimmed();
```
Don't understand the logic behind silently trimming in both directions

fmatthew5876 replied:
I guess its to silently fix whitespace misktakes made by users either in the config file or in the UI. This was suggested by an earlier reviewer. I'm fine either way, I just want qbit to work in my FreeBSD jail!

glassez replied:
>fix whitespace misktakes made by users either in the config file or in the UI.

Yes, it is.

Chocobo1 commented:
@fmatthew5876 
Thank you!

ciscler commented:
Hi all,
I am runnning qbittorrent v3.3.16 in FreeBSD 11.1-RELEASE-p9 Jail. But Server is not listening on the webgui port even I have set WebU\Address=* in configuration.

[Preferences]
WebUI\Address=*
WebUI\Port=8080

qBittorentlog:

`(N)` 2018-04-17T15:16:46 - qBittorrent v3.3.16 started
(I) 2018-04-17T15:16:46 - qBittorrent is trying to listen on any interface port: 8999
(I) 2018-04-17T15:16:46 - qBittorrent is trying to listen on any interface port: 8999
(N) 2018-04-17T15:16:46 - Peer ID: -qB33G0-
(N) 2018-04-17T15:16:46 - Peer ID: -qB33G0-
(N) 2018-04-17T15:16:46 - HTTP User-Agent is 'qBittorrent/3.3.16'
(N) 2018-04-17T15:16:46 - HTTP User-Agent is 'qBittorrent/3.3.16'
(I) 2018-04-17T15:16:46 - DHT support [ON]
(I) 2018-04-17T15:16:46 - DHT support [ON]
(I) 2018-04-17T15:16:46 - Local Peer Discovery support [ON]
(I) 2018-04-17T15:16:46 - Local Peer Discovery support [ON]
(I) 2018-04-17T15:16:46 - PeX support [ON]
(I) 2018-04-17T15:16:46 - PeX support [ON]
(I) 2018-04-17T15:16:46 - Anonymous mode [OFF]
(I) 2018-04-17T15:16:46 - Anonymous mode [OFF]
(I) 2018-04-17T15:16:46 - Encryption support [ON]
(I) 2018-04-17T15:16:46 - Encryption support [ON]
(I) 2018-04-17T15:16:46 - Embedded Tracker [OFF]
(I) 2018-04-17T15:16:46 - Embedded Tracker [OFF]
(I) 2018-04-17T15:16:46 - UPnP / NAT-PMP support [ON]
(I) 2018-04-17T15:16:46 - UPnP / NAT-PMP support [ON]
(I) 2018-04-17T15:16:46 - GeoIP database loaded. Type: GeoLite2-Country. Build time: Wed Apr 4 00:40:13 2018.
(I) 2018-04-17T15:16:46 - GeoIP database loaded. Type: GeoLite2-Country. Build time: Wed Apr 4 00:40:13 2018.
(C) 2018-04-17T15:16:46 - Web UI: Unable to bind to port 8080
(C) 2018-04-17T15:16:46 - Web UI: Unable to bind to port 8080
(C) 2018-04-17T15:16:46 - qBittorrent failed listening on interface 0.0.0.0 port: TCP/8999. Reason: Protocol not supported.
(C) 2018-04-17T15:16:46 - qBittorrent failed listening on interface 0.0.0.0 port: TCP/8999. Reason: Protocol not supported.
(I) 2018-04-17T15:16:46 - qBittorrent is successfully listening on interface 192.168.0.96 port: TCP/8999
(I) 2018-04-17T15:16:46 - qBittorrent is successfully listening on interface 192.168.0.96 port: TCP/8999
(I) 2018-04-17T15:16:46 - qBittorrent is successfully listening on interface 0.0.0.0 port: UDP/8999
(I) 2018-04-17T15:16:46 - qBittorrent is successfully listening on interface 0.0.0.0 port: `UDP/8999`

Has anyone an idea?

thalieht commented:
This PR is not in 3.3.16

ciscler commented:
And is there a way to get it running?

thalieht commented:
I don't know if 3.3.x is still supported.

ciscler commented:
So there is no way to get this running with freeBSD? The latest Version in FreeBSD is 3.3.16
